dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java

```diff
@@ -0,0 +1,174 @@
+     *
+     * @param workerGroup the identifier for the worker group, used to distinguish different task queues
+     * @param taskExecutionRunnable an instance of ITaskExecutionRunnable representing the task to be executed
+     * @param delayTimeMills the delay time before the task is executed, in milliseconds
+     */
+    public synchronized boolean addTaskToWorkerGroup(String workerGroup, ITaskExecutionRunnable taskExecutionRunnable,
+                                                     long delayTimeMills) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            if (workerGroupTaskDispatcher.checkCloseDispatchWorkerComplete()) {
+                dispatchWorkerMap.remove(workerGroup);
+                return false;
+            }
+            return workerGroupTaskDispatcher.addTaskToWorkerGroupQueue(taskExecutionRunnable, delayTimeMills);
+        } else {
+            log.error("workerGroupTaskDispatcher {} not found, will set task {} fail",
+                    workerGroup, taskExecutionRunnable.getTaskInstance().getId());
+        }
+        return false;
+    }
+
+    /**
+     * Stops a specific worker group's task dispatch waiting queue looper.
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    private synchronized void deleteWorkerGroup(String workerGroup) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            workerGroupTaskDispatcher.markDispatcherClosing();
+        } else {
+            log.warn("workerGroupTaskDispatcher {} not found", workerGroup);
+        }
+    }
+
+    /**
+     * add workerGroup
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    private synchronized void addWorkerGroup(String workerGroup) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher == null) {
+            workerGroupTaskDispatcher = new WorkerGroupTaskDispatcher(workerGroup, taskExecutorClient);
+            dispatchWorkerMap.put(workerGroup, workerGroupTaskDispatcher);
+            workerGroupTaskDispatcher.start();
+        } else {
+            workerGroupTaskDispatcher.markDispatcherStart();
+        }
+    }
+
+    /**
+     * Stop all workerGroupTaskDispatchWaitingQueueLooper
+     */
+    @Override
+    public void close() throws Exception {
+        log.info("WorkerGroupTaskDispatcherManager start close");
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            try {
+                entry.getValue().markDispatcherClosing();
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
+            }
+        }
+        log.info("WorkerGroupTaskDispatcherManager closed");
```

Comment:
In fact you should closed the dispatcher here not just closing.

Reply:
change to 
```
 entry.getValue().close();
```
WorkerGroupTaskDispatcher will be left as a TODO.